gramineproject/gramine
# [tools] Fix OID for RA-TLS

The oid parameter of mbedtls_x509write_crt_set_extension() accepts an
encoded OID octet string, rather than an ASN.1 encoded OID object (type 6).

Without the fix, the SGX quote extension in the dump result shows a
weird OID:
            0.6.9.42.840.113741.1337.6:
                ..............r3..L..

The corrected OID should be:
            1.2.840.113741.1337.6:
                ..............r3..L..

Signed-off-by: Jia Zhang <zhang.jia@linux.alibaba.com>



## Description of the changes 



The OID should be canonical and defined correctly. Currently the Evidence Extension format has a wrong OID due to misusing the OID value.

## How to test this PR? 

```shell
cd CI-Examples/ra-tls-mbedtls
vim src/client.c
```

manually apply this patch:
```C
diff --git a/CI-Examples/ra-tls-mbedtls/src/client.c b/CI-Examples/ra-tls-mbedtls/src/client.c
index 4d2f71ba..1655dad2 100644
--- a/CI-Examples/ra-tls-mbedtls/src/client.c
+++ b/CI-Examples/ra-tls-mbedtls/src/client.c
@@ -108,6 +108,10 @@ static int my_verify_measurements(const char* mrenclave, const char* mrsigner,
 static int my_verify_callback(void* data, mbedtls_x509_crt* crt, int depth, uint32_t* flags) {
     (void)data;
 
+    FILE *fp = fopen("/tmp/dump.der", "wb");
+    fwrite(crt->raw.p, crt->raw.len, 1, fp);
+    fclose(fp);
+
     if (depth != 0) {
         /* the cert chain in RA-TLS consists of single self-signed cert, so we expect depth 0 */
         return MBEDTLS_ERR_X509_INVALID_FORMAT;
```

Then dump the certificate and show its content in text:
```shell
make app dcap RA_TYPE=dcap
gramine-sgx ./server &
RA_TLS_ALLOW_DEBUG_ENCLAVE_INSECURE=1 \
RA_TLS_ALLOW_OUTDATED_TCB_INSECURE=1 \
./client dcap
openssl x509 -in /tmp/dump.der -inform der -text | grep 'X509v3 extensions' -A9
```

Results is:
```
        X509v3 extensions:
            X509v3 Basic Constraints: 
                CA:FALSE
            X509v3 Subject Key Identifier: 
                65:23:59:F3:28:2A:5A:DD:79:4E:18:85:48:E6:97:38:73:E6:F7:D6
            X509v3 Authority Key Identifier: 
                keyid:65:23:59:F3:28:2A:5A:DD:79:4E:18:85:48:E6:97:38:73:E6:F7:D6

            0.6.9.42.840.113741.1337.6: 
                ..............r3..L..
```

After applying this patch, the result would be:
```
        X509v3 extensions:
            X509v3 Basic Constraints: 
                CA:FALSE
            X509v3 Subject Key Identifier: 
                3A:09:B0:A2:FA:6C:A8:55:BF:AA:0F:67:74:E9:18:08:E8:B5:2A:E5
            X509v3 Authority Key Identifier: 
                keyid:3A:09:B0:A2:FA:6C:A8:55:BF:AA:0F:67:74:E9:18:08:E8:B5:2A:E5

            1.2.840.113741.1337.6: 
                ..............r3..L..
```


---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/gramineproject/gramine/805)
